### DISCUSSION TITLE
custom view on node element on GLSP editor

### DISCUSSION THREAD

**Discussion by Abhay0294** [eclipse-glsp/glsp#811] [Q&A] [answered]
Hi,

I want to implement the following requirement,

1. When user clicks on the node he should get that custom red icon(svg element) view icon over the node.
2. when he clicks on that I should able to fire some custom action.(Basically want to implement rotate operation of the node).

Also attached the sanp of requirement for reference

<img width="439" alt="requiremnt1" src="https://user-images.githubusercontent.com/57217189/206655261-973a4f52-0548-4081-91be-1d4ec5bace28.png">

Requesting suggestions for the same

**Comment by martin-fleck-at** [comment, marked as answer]
Hi @Abhay0294, That sounds like an interesting idea! So what you would need is a mouse listener that listens to clicks on an element and then display some icon over the node and react to clicking on that node, correct? Typically, we encapsulate such a tool that is registered. Tools can then be enabled and disabled either by default or through a dedicated `EnableToolsAction`.

I believe the functionality can be compared to the resize handles that we display once you click on an element that has the resize features and the necessary bounds. The resize handles are implemented as part of the [ChangeBoundsTool](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/tools/change-bounds-tool.ts), a default tool that takes care of resizing and moving of elements by applying client-side feedback first and then after the user action has finished, sends an update to the server. I'd suggest you have a look at that but basically what happens is this. 
1. When the tool is enabled (default), it installs a mouse listener which is uninstalled when the tool is disabled. This is something that usually prevents conflicts with other tools/listeners.
2. The mouse listener listens to a user click/mouse down and if the element matches the requirements (can be rotated in your case), then we dispatch a feedback action that displays the resize handles on that element. In our case, we actually add the elements to the model, in your case, if you only want to show a simple button, you could also use a [UI extension](https://www.eclipse.org/glsp/documentation/ui-extensions/) that handles any further behavior.
3. If the user clicks on a resize handle and starts dragging, we create actions on the client that ensure that the element is properly resized (rotated in your case) to ensure the user actually sees what he is doing. If you only require a click functionality, even better, then you can trigger your server-side function now.
4. If the user finishes dragging we send an update with the new size to the server.

I hope this will help you get started on this.

**Reply by Abhay0294** [reply]
Thanks @martin-fleck-at

**Reply by Abhay0294** [reply]
Hello @martin-fleck-at ,

I am able to achieve that visual implementation on the node i.e. implementing node rotator handle and when user clicks on that the node rotation takes place.

https://user-images.githubusercontent.com/57217189/211131948-67a7b56b-eab4-4a8d-9e3b-ba1b3b79149b.mp4

But view wise it doesn't seems to be rotation rather it looks like resizing(attached video of the actual implementation above).

Below is the code for the rotation logic:

`function changedNodeBounds(element: BoundsAwareModelElement): ElementAndBounds {

    let newX = 0;
    let newY = 0;

    if (element.bounds.x >= 0) {
        newX = element.bounds.x + (element.bounds.width / 2) - (element.bounds.height / 2);
    }
    else {
        newX = element.bounds.x + (element.bounds.width / 2) - (element.bounds.height / 2);
    }

    if (element.bounds.y >= 0) {
        newY = element.bounds.y - (element.bounds.width / 2) + (element.bounds.height / 2);
    }
    else {
        newY = element.bounds.y - (element.bounds.width / 2) + (element.bounds.height / 2);
    }

    return {
        elementId: element.id,
        newPosition: {
            x: newX,
            y: newY
        },
        newSize: {
            width: element.bounds.height,
            height: element.bounds.width
        }
    };
} 

export class AddRotationFeature extends MouseListener {

    @inject(GLSPActionDispatcher) protected readonly action: GLSPActionDispatcher;

    override mouseDown(target: SModelElement & BoundsAware, event: MouseEvent): (Action | Promise<Action>)[] {
        if (target instanceof RotatorElement) {
            const parentElement: any = target.parent;
            const changeBoundsAction = ChangeBoundsOperation.create([changedNodeBounds(parentElement)]);
            setTimeout(() => this.action.dispatch(changeBoundsAction), 0);
        }
        return [];
    }
}`

I am dispatching `changeBoundsAction ` on the click of rotator handle on the node.

Requesting for the improvements in this logic so I can achieve the actual rotation view wise. Thanks

**Reply by martin-fleck-at** [reply]
Hi @Abhay0294, if I understand you correctly you want to have a different animation than the default resize animation when using rotation, is that correct?

We do not use any custom animations so far in GLSP but all Sprotty animations are available by default: [ResizeAnimation](https://github.com/eclipse/sprotty/blob/master/packages/sprotty/src/features/bounds/resize.ts#L29), [FadeAnimation](https://github.com/eclipse/sprotty/blob/master/packages/sprotty/src/features/fade/fade.ts#L31), [MoveAnimation](https://github.com/eclipse/sprotty/blob/master/packages/sprotty/src/features/move/move.ts#L265), [MorphEdgesAnimation](https://github.com/eclipse/sprotty/blob/master/packages/sprotty/src/features/move/move.ts#L298), and [ViewportAnimation](https://github.com/eclipse/sprotty/blob/master/packages/sprotty/src/features/viewport/viewport.ts#L144). 

Each animation needs to calculate the state of the model at a certain point within the animation duration given by the command execution context. Animations are typically calculated "globally" [between two model states during the model update](https://github.com/eclipse/sprotty/blob/e898c14efbb54155049c2aa6c14dd5ead150eaa1/packages/sprotty/src/features/update/update-model.ts#L278) but there is a case where the animations are executed "directly" [during the move command](https://github.com/eclipse/sprotty/blob/master/packages/sprotty/src/features/move/move.ts#L138) execution if the MoveAction has the animate flag set to true, see here. Without having tried it, I think you might be able to do something similar, i.e., create your own custom animation and then apply it when handling your action but I have never done it myself.

**Reply by Abhay0294** [reply]
Thanks @martin-fleck-at , will try to implement the custom animation.

One more thing to ask, recently I noticed a bug in the implementation(also attached a video below of the same).

https://user-images.githubusercontent.com/57217189/211300960-ab3ada5f-f9b8-43cb-a0c2-e99b9470347c.mp4

From the video above I want to show that abnormal behavior in the rotation action. So what is happening:

1. When I click on the node rotator handle to do the rotation, sometimes it tries to change the dimension but again it comes to original dimension.
2. Sometimes when it is performing the rotation it switches dimensions(between original and new dimension) multiple times, which we can see in the video.

I have already attached the code snippet of the implementation above in this thread.

Requesting for your suggestion in this to know what could be the issue and the probable solutions to fix this.

**Reply by martin-fleck-at** [reply]
Hmm... without actually debugging the application it is very hard to tell. It might be that the mouseDown is triggered twice in some cases. However, what strikes me as a bit odd is the fact that you dispatch the ChangeBoundsOperation after some timeout here: `setTimeout(() => this.action.dispatch(changeBoundsAction), 0);`. Is there any particular reason you do not simply return it as part of the function, e.g., `return [changeBoundsAction]`?

**Reply by Abhay0294** [reply]
Hi @martin-fleck-at ,

That issue was there before I added  `setTimeout(() => this.action.dispatch(changeBoundsAction), 0);`. 

Previously what I was doing
`export class AddRotationFeature extends MouseListener {

    @inject(GLSPActionDispatcher) protected readonly action: GLSPActionDispatcher;

    override mouseDown(target: SModelElement & BoundsAware, event: MouseEvent): (Action | Promise<Action>)[] {
        if (target instanceof RotatorElement) {
            const parentElement: any = target.parent;
            const changeBoundsAction = ChangeBoundsOperation.create([changedNodeBounds(parentElement)]);
            this.action.dispatch(changeBoundsAction);
        }
        return [];
    }
}`

At this point also I am facing the same issue. Added that timeout due to a thought that there might be some async operation taking place.

**Reply by martin-fleck-at** [reply]
Could you not simply return the action directly?

```
export class AddRotationFeature extends MouseListener {
 @inject(GLSPActionDispatcher) protected readonly action: GLSPActionDispatcher;

 override mouseDown(target: SModelElement & BoundsAware, event: MouseEvent): (Action | Promise<Action>)[] {
    if (target instanceof RotatorElement) {
        const parentElement: any = target.parent;
        const changeBoundsAction = ChangeBoundsOperation.create([changedNodeBounds(parentElement)]);
        return [changeBoundsAction];
    }
    return [];
 }
}
```

**Reply by Abhay0294** [reply]
Tried this too @martin-fleck-at and tested but still getting the same issue

**Reply by Abhay0294** [reply]
@martin-fleck-at  just for the info below are my bindings for this node rotation implementation:

```
configureCommand({ bind, isBound }, RotatorHandle.commands.DrawRotatorPointCommand);
    configureCommand({ bind, isBound }, RotatorHandle.commands.RemoveRotatorPointCommand);
    configureView({ bind, isBound }, RotatorHandle.model.RotatorElement.TYPE, RotatorHandle.view);
    bind(RotatorHandle.tool).toSelf().inSingletonScope();
    bind(TYPES.IDefaultTool).to(RotatorHandle.tool);
    bind(RotatorHandle.addRotation).toSelf().inSingletonScope();
    bind(TYPES.MouseListener).to(RotatorHandle.addRotation);
```

**Reply by martin-fleck-at** [reply]
@Abhay0294 Just from the code, everything looks good to me. The only thing I maybe would suggest is to move the mouse listener for the rotator handle inside the rotator handle tool, i.e., register the mouse listener when the tool gets enabled and de-register it when the tool gets disabled, e.g., as in the [edit label tool](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/edit-label/edit-label-tool.ts). It probably does not make a lot of difference in your case since you are checking for a very specific element but the idea behind this is to avoid conflicts between mouse listeners of different tools as much as possible.

Nevertheless, I don't see anything from the code you posted that should cause this behavior. When you debug your mouse listener, do you see whether the method gets actually called twice or none at all? Or is there some problem with the update of the change bounds operation?

**Reply by Abhay0294** [reply]
Hello @martin-fleck-at ,
I found the root cause for that abnormal behaviour. The hover action on the node is causing that behaviour. I have attached the detailed video below:

https://user-images.githubusercontent.com/57217189/211780840-09bbcbb9-6a75-483b-b2a9-1c56cad07ec8.mp4

So from the video we can see that:
1. Rotation is working fine unless and until we stop for sometime(by hovering on the node) to make the tooltip(in this case joinstate) to appear.
2. It starts misbehaving when the tooltip is appeared and after that when rotator handle is clicked.

So what I did, I just removed that tooltip on hover and then tested and it worked fine.

So now wondering how can I adjust both the things i.e. tooltip should be there with no bug in rotation. 
Basically the issue is the multiple event interference.
Your inputs in this case will be highly appreciated.

**Reply by martin-fleck-at** [reply]
@Abhay0294 Wow, that looks really interesting! Not sure why this is happening and I never experienced this myself but I'll try to reproduce this tomorrow or maybe, if you have some time, you can set up a minimal example based on the public GLSP workflow example.

**Reply by Abhay0294** [reply]
Hello @martin-fleck-at 
The issue is resolved now, and the fix was just one liner for this weird behavior.
So what I did, I just replaced the title tag for the rotator handle SVG component as <title></title> which is empty and on now on hover it doesn't show anything and now the rotation implementation is working fine.

Thanks a lot @martin-fleck-at , your inputs really helped me in my R&D work during this implementation.

**Reply by martin-fleck-at** [reply]
@Abhay0294 That's great news! And thank you very much for sharing your solution so we can keep an eye out for similar problems.
